add addressindex RPC for bitcore block explorer

LarryRuane opened the pull request:
Addresses #3708

LarryRuane commented:
Please only look at the commits (currently only one, the latest) that are unique to this PR (since this PR is layer on others).

mms710 commented:
Note for reviewers: these PRs must be reviewed and merged in this order:
#3863 
#3926 
#3953 
#3972 

Between every merge, ping Larry to rebase the next PR in the list so you can review and merge the next one down the line

Eirik0 commented on src/txmempool.h:
```diff
@@ -152,6 +153,15 @@ class CTxMemPool
     indexed_transaction_set mapTx;
+
+private:
+    typedef std::map<CMempoolAddressDeltaKey, CMempoolAddressDelta, CMempoolAddressDeltaKeyCompare> addressDeltaMap;
```
Do we end up using these typedefs more later? I personally prefer having the raw types. It's convenient not to have to type all this when we use these in `txmempool.cpp`, but in those cases it looks like we can get around that with the `auto` keyword.

LarryRuane replied:
@Eirik0, good suggestion, fixed in 18d9594ce7106403abf6119fb204064acbe05c15. I also improved a bunch of other stuff (force-pushed), so please take at least a quick look at everything again.

Eirik0 commented:
@zkbot try

mdr0id commented:
@zkbot retry

Might have got in odd state from cred updates, so retrying

str4d commented:
I manually cancelled the build because I saw that one of the builders had transiently failed early on. Waiting a bit for the workers to be torn down, and then we can retry again.

str4d commented:
@zkbot retry

rex4539 commented:
Shouldn't all references to "satoshis" in the code, be "zatoshis" instead?

LarryRuane commented:
@rex4539, the reason for `satoshis` is that the modified-for-zcash Insight block explorer (in a few different repos within https://github.com/zcash-hackworks) references `satoshis` and not `zatoshis`. That's because this block explorer comes from Bitcoin. That's why the upstream source of this PR, https://github.com/zcash-hackworks/zcash-patched-for-explorer, similarly references `satoshis`, for example https://github.com/zcash-hackworks/zcash-patched-for-explorer/blob/v2.0.4-insight-explorer/src/rpc/misc.cpp#L624 (and other places in that file). 

When this and the related PRs are merged into `zcash/zcash master`, we'd like the resulting `zcashd` to be a drop-in replacement for the hackworks version (i.e. no need to change the Insight explorer code), which requires us to be backward compatible with the hackworks version.

I considered adding `zatoshis` to the JSON output (have both), so that later we could transition the Insight code to reference that, and then eventually remove `satoshis`, but it didn't seem worth it to me (let me know if you disagree); end-users don't ever see these JSON  objects.

LarryRuane commented:
The latest commit adds testing of `DisconnectBlock()`. This function is called on a reorg, and should reverse all the effects of an earlier `ConnectBlock()`. There's Insight Explorer code in both functions to adjust the `addressindex`, but until now, we weren't testing the `DisconnectBlock()` changes. I manually tested the latest commit by making sure that the RPC test `addressindex.py` fails if the addressindex code in `DisconnectBlock()` is disabled. Before the latest commit, that RPC test would succeed even if the code in `DisconnectBlock()` is disabled.

I'll squash after reviews.

Eirik0 commented on qa/rpc-tests/addressindex.py:
```diff
@@ -0,0 +1,389 @@
+
+        # Test DisconnectBlock() by invalidating the most recent mined block
+        tip = self.nodes[1].getchaintips()[0]
+        for i in range(3):
```
This could be changed to 

```
for node in self.nodes:
    ...
```

Eirik0 commented on qa/rpc-tests/addressindex.py:
```diff
@@ -0,0 +1,389 @@
+        change = change_vout[0]['scriptPubKey']['addresses'][0]
+        bal = self.nodes[2].getaddressbalance(change)
+        assert(bal['received'] > 0)
+        assert(bal['received'] < (4-3)*COIN)
```
Maybe add a comment here about why we do not check the exact amount.

LarryRuane replied:
@eirik0, this change value is not consistent across multiple runs of this test. Usually, the balance is 99999774 zatoshis, but about one of 5 times, it's 99999775. Is that expected? Is there some randomness in the transaction fee? But at least for now, I'll add a comment, as you suggest.

Eirik0 commented on qa/rpc-tests/addressindex.py:
```diff
@@ -0,0 +1,389 @@
+        for i in (0, 2):    # address b listed in args first and last
+            assert_equal(mempool[i]['address'], b)
+            assert_equal(mempool[i]['satoshis'], 3*COIN)
+            assert_equal(mempool[i]['txid'], txid)
```
This could become:
```
        mempool = self.nodes[0].getaddressmempool({'addresses': [addr2, addr1, addr2]})
        assert_equal(len(mempool), 3)
        [assert_equal(tx['txid'], txid) for tx in mempool]
        # addr2 listed in args first and last
        assert_equal(mempool[0]['address'], addr2)
        assert_equal(mempool[0]['satoshis'], 3 * COIN)
        assert_equal(mempool[1]['address'], addr1)
        assert_equal(mempool[1]['satoshis'], (-4) * COIN)
        assert_equal(mempool[2]['address'], addr2)
        assert_equal(mempool[2]['satoshis'], 3 * COIN)
```

If we wanted to eliminate the loop (which I found marginally confusing).

Eirik0 commented on qa/rpc-tests/addressindex.py:
```diff
@@ -0,0 +1,389 @@
+            'addresses': [a],
+        })
+        assert_equal(len(deltas), len(expected_deltas))
+        for z in zip(deltas, expected_deltas):
```
Love the zip!

However, I think this would be more readable as:
```
        for i in range(len(deltas)):
            assert_equal(deltas[i]['address'],   a)
            assert_equal(deltas[i]['height'],    expected_deltas[i]['height'])
            ....
```

Eirik0 commented on qa/rpc-tests/addressindex.py:
```diff
@@ -0,0 +1,389 @@
+            assert_equal(z[0]['satoshis'],  z[1]['satoshis'])
+            assert_equal(z[0]['txid'],      z[1]['txid'])
+
+        deltas_limited = self.nodes[1].getaddressdeltas({
```
If we wanted to be able to do this in a single line, we could add the following function:
```
def getaddressdeltas(node, addresses, start, end, chainInfo=None):
    params = {
        'addresses': addresses,
        'start': start,
        'end': end,
    }
    if chainInfo is not None:
        params.update({'chainInfo': chainInfo})
    return node.getaddressdeltas(params)
```

Eirik0 commented on qa/rpc-tests/addressindex.py:
```diff
@@ -0,0 +1,389 @@
+        assert_equal(self.nodes[1].getaddresstxids(addr), [txid])
+        bal = self.nodes[2].getaddressbalance(addr)
+        assert_equal(bal['balance'], 3*COIN)
+        assert_equal(bal['received'], 3*COIN)
```
General nit: There are not spaces surrounding the +,-,* operators.

LarryRuane commented on qa/rpc-tests/addressindex.py:
```diff
@@ -0,0 +1,389 @@
+
+        # only the oldest 5 transactions are in the unspent list
+        height_txids = self.nodes[1].getaddresstxids({
+            'addresses': [addr_p2pkh, addr_p2pkh],
```
Yes, just to test how dups are handled. I'll add a comment to clarify.

LarryRuane commented on qa/rpc-tests/addressindex.py:
```diff
@@ -0,0 +1,389 @@
+    start_nodes,
+    stop_nodes,
+    connect_nodes,
+    connect_nodes_bi,
```
@Eirik0, thanks; how did you notice this, just by inspection? Or did you run a tool, and if so, which one (so I can prevent these kinds of mistakes in the future).

LarryRuane commented:
@Eirik0, addressed your excellent review comments on the test, thanks.

Eirik0 commented on src/rpc/misc.cpp:
```diff
@@ -483,6 +484,447 @@ UniValue setmocktime(const UniValue& params, bool fHelp)
+    if (fHelp || params.size() != 1)
+        throw runtime_error(
+            "getaddressbalance\n"
+            "\nReturns the balance for an address(es) (requires insightexplorer to be enabled).\n"
```
In other places I don't think we have `(es)`.

Eirik0 commented on src/rpc/misc.cpp:
```diff
@@ -483,6 +484,447 @@ UniValue setmocktime(const UniValue& params, bool fHelp)
+    int end = 0;
+    addressesHeightRange(params, addresses, addressIndex, start, end);
+
+    std::set<std::pair<int, std::string> > txids;
```
Does this need to be a pair? It seems like we are adding the height to the set, but then this is not used in the result.

Eirik0 commented on src/rpc/misc.cpp:
```diff
@@ -483,6 +484,447 @@ UniValue setmocktime(const UniValue& params, bool fHelp)
+
+    std::set<std::pair<int, std::string> > txids;
+    UniValue result(UniValue::VARR);
+    for (const auto& it : addressIndex) {
```
This loop and subsequent if statement confuse me. Why not add the txids directly to the result? Why do we need `if (addresses.size() > 1)`?

LarryRuane replied:
The loop is okay (we want to return all transactions that include any of the given addresses), but the `if` is over-optimized (if it's an optimization at all). I'll simplify.

Eirik0 commented on src/txmempool.cpp:
```diff
@@ -121,6 +121,66 @@ bool CTxMemPool::addUnchecked(const uint256& hash, const CTxMemPoolEntry &entry,
+    mapAddressInserted.insert(make_pair(txhash, inserted));
+}
+
+void CTxMemPool::getAddressIndex(std::vector<std::pair<uint160, int> > &addresses,
```
Should `&addresses` be const?

Eirik0 commented on src/rpc/misc.cpp:
```diff
@@ -483,6 +484,447 @@ UniValue setmocktime(const UniValue& params, bool fHelp)
+
+    for (const auto& it : indexes) {
+        std::string address;
+        if (!getAddressFromIndex(it.first.type, it.first.addressBytes, address)) {
```
It would be nice if there were another way to do this. We had the addresses when we looked up the indexes, but lost their associations when calling `mempool.getAddressIndex(...)`. Could this function be changed to populate a std:map<std::string, std::pair<...>> rather than a std::vector<std:pair<...>> to avoid having to do this?

Eirik0 replied:
I realize that a map may be inconvenient because of the sorting. I am fine with leaving this and potentially trying to come up with something more creative later.

LarryRuane commented on src/main.cpp:
```diff
@@ -1615,6 +1619,31 @@ bool GetSpentIndex(CSpentIndexKey &key, CSpentIndexValue &value)
     return pblocktree->ReadSpentIndex(key, value);
 }
 
+bool GetAddressIndex(uint160 addressHash, int type,
```
Makes sense for nontrivial types like `uint160` and strings, but not for simple types like `int` (it's less efficient to pass an 8-byte pointer to an `int` than a 4-byte `int`). In this case, I'll fix the `uint160 addressHash`.

LarryRuane commented on src/rpc/misc.cpp:
```diff
@@ -483,6 +484,447 @@ UniValue setmocktime(const UniValue& params, bool fHelp)
+            throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY,
+                "Addresses is expected to be an array");
+        }
+    } else {
```
Very clever idea, but I'm not going to do this. It works in the python test, but I can't get it to work using a live node:
```
$ src/zcash-cli getaddressbalance '["tmYXBYJj1K7vhejSec5osXK2QsGa5MTisUQ", "tmTzyyT7PDiAfhx7V53kvtbnU1SKCv3niiz"]'
error code: -5
error message:
Invalid address
$ 
```
(I tried a few variations with no success.) I don't think it's worth figuring this out, and also, as you said, there would be documentation work to do too.

Eirik0 replied:
Sounds good. Thanks for looking in to this!

LarryRuane commented on src/rpc/misc.cpp:
```diff
@@ -483,6 +484,447 @@ UniValue setmocktime(const UniValue& params, bool fHelp)
+    bool includeChainInfo = false;
+    if (params[0].isObject()) {
+        UniValue chainInfo = find_value(params[0].get_obj(), "chainInfo");
+        if (chainInfo.isBool()) {
```
Can't just remove the `if` statement, because then it becomes a required argument. But I fixed it (`get_bool()` will throw if not a boolean):
```
    if (params[0].isObject()) {
        UniValue chainInfo = find_value(params[0].get_obj(), "chainInfo");
        if (!chainInfo.isNull()) {
            includeChainInfo = chainInfo.get_bool();
        }
    }
```

Eirik0 replied:
Sounds good. I think when I made this comment I hadn't realized this was an optional argument yet. If you were going through and updating the documentation, you could change this too.

`"  \"chainInfo\"  (boolean) Include chain info with results\n"`
->
`"  \"chainInfo\"  (boolean, optional, default=false) Include chain info with results\n"`

LarryRuane commented on src/rpc/misc.cpp:
```diff
@@ -483,6 +484,447 @@ UniValue setmocktime(const UniValue& params, bool fHelp)
+    if (params[0].isObject()) {
+        UniValue startValue = find_value(params[0].get_obj(), "start");
+        UniValue endValue = find_value(params[0].get_obj(), "end");
+        if (startValue.isNum() && endValue.isNum()) {
```
Another good catch, I'll fix it similarly to `chainInfo` by changing the `if` to:
```
        if (!startValue.isNull() && !endValue.isNull()) {
```
(Then the call to `get_int()` will throw if not an integer.) I hope these changes don't break the existing Insight block explorer. That would be a very bad design on its part if it does!

Eirik0 replied:
Hmm. I guess this one is a little trickier. It seems to be the case that if you have a start value that you must have an end value and vice versa. Maybe this should be:

```
if (!startValue.isNull() && !endValue.isNull()) {
    // get_int()
} else if (startValue.isNull() != endValue.isNull()) { // either both values must be null or neither
    // throw error
}
```

or something like that.

LarryRuane commented on src/rpc/misc.cpp:
```diff
@@ -483,6 +484,447 @@ UniValue setmocktime(const UniValue& params, bool fHelp)
+            "      ,...\n"
+            "    ],\n"
+            "  \"chainInfo\"  (boolean) Include chain info with results\n"
+            "}\n"
```
I am inclined against changing this; as you say, it may break existing insight explorer code.

Eirik0 replied:
Yes, let's not worry about this for the time being.

Eirik0 commented on src/rpc/misc.cpp:
```diff
@@ -483,6 +484,447 @@ UniValue setmocktime(const UniValue& params, bool fHelp)
+// Parse an address list and optional start and end height values,
+// and then fetch the corresponding addressindex information.
+static void addressesHeightRange(
+    const UniValue& params,
```
If this is at all a pain, I am happy for us to come back to this later.

Eirik0 commented on src/rpc/misc.cpp:
```diff
@@ -483,6 +484,459 @@ UniValue setmocktime(const UniValue& params, bool fHelp)
+            "\nReturns all mempool deltas for an address (requires insightexplorer to be enabled).\n"
+            "\nArguments:\n"
+            "{\n"
+            "  \"addresses\"\n"
```
```suggestion
            "  \"addresses\":\n"
```

Eirik0 commented on src/rpc/misc.cpp:
```diff
@@ -483,6 +484,459 @@ UniValue setmocktime(const UniValue& params, bool fHelp)
+            "\nReturns the txids for given transparent addresses (requires insightexplorer to be enabled).\n"
+            "\nArguments:\n"
+            "{\n"
+            "  \"addresses\"\n"
```
```suggestion
            "  \"addresses\":\n"
```

LarryRuane replied:
Thanks, fixed your latest suggestions, force-pushed

str4d commented on qa/rpc-tests/addressindex.py:
```diff
@@ -0,0 +1,363 @@
+        self.sync_all()
+
+        # the one tx in the mempool refers to addresses addr1 and addr2
+        mempool = self.nodes[0].getaddressmempool({'addresses': [addr2, addr1, addr2]})
```
Is it significant to the test that `addr2` is duplicated?

LarryRuane replied:
Yes, this was intentional, I wanted to test that the results are returned in the order corresponding to the arguments, including duplicate arguments. I will add a comment to clarify.

LarryRuane commented on src/main.cpp:
```diff
@@ -1615,6 +1619,31 @@ bool GetSpentIndex(CSpentIndexKey &key, CSpentIndexValue &value)
 }
 
+bool GetAddressIndex(const uint160& addressHash, int type,
+                     std::vector<CAddressIndexDbEntry>& addressIndex,
```
This typedef is defined here: https://github.com/zcash/zcash/commit/6cf0e50b56b39eda3cd2cf91e99a594f36e01618#diff-e3b0b9197f4a56f9e534ffafe96fd306R29 which is part of https://github.com/zcash/zcash/pull/3837.

LarryRuane commented on src/main.cpp:
```diff
@@ -1615,6 +1619,31 @@ bool GetSpentIndex(CSpentIndexKey &key, CSpentIndexValue &value)
+}
+
+bool GetAddressUnspent(const uint160& addressHash, int type,
+                       std::vector<CAddressUnspentDbEntry>& unspentOutputs)
```
That was defined here: https://github.com/zcash/zcash/commit/6cf0e50b56b39eda3cd2cf91e99a594f36e01618#diff-e3b0b9197f4a56f9e534ffafe96fd306R28, also part of https://github.com/zcash/zcash/pull/3837.

LarryRuane commented on src/txdb.cpp:
```diff
@@ -326,7 +326,7 @@ bool CBlockTreeDB::ReadAddressUnspentIndex(uint160 addressHash, int type, std::v
             break;
         CAddressUnspentValue nValue;
-        if (pcursor->GetValue(nValue))
+        if (!pcursor->GetValue(nValue))
```
This is the only one, I checked carefully.

LarryRuane commented on src/rpc/misc.cpp:
```diff
@@ -483,6 +484,459 @@ UniValue setmocktime(const UniValue& params, bool fHelp)
+        UniValue startValue = find_value(params[0].get_obj(), "start");
+        UniValue endValue = find_value(params[0].get_obj(), "end");
+        // If either is not specified, the other is ignored.
+        if (!startValue.isNull() && !endValue.isNull()) {
```
Answered earlier (throw if incorrect data type).

LarryRuane commented on src/rpc/misc.cpp:
```diff
@@ -483,6 +484,459 @@ UniValue setmocktime(const UniValue& params, bool fHelp)
+            "      ,...\n"
+            "    ]\n"
+            "  \"start\" (number, optional) The start block height\n"
+            "  \"end\" (number, optional) The end block height\n"
```
There do not seem to be any other such RPCs.

LarryRuane commented on src/rpc/rawtransaction.cpp:
```diff
@@ -66,7 +66,9 @@ UniValue TxJoinSplitToJSON(const CTransaction& tx) {
         UniValue joinsplit(UniValue::VOBJ);
 
         joinsplit.push_back(Pair("vpub_old", ValueFromAmount(jsdescription.vpub_old)));
+        joinsplit.push_back(Pair("vpub_oldZat", jsdescription.vpub_old));
```
Fix problem reported here https://github.com/zcash/zcash/pull/3972#issuecomment-513510051 (but wanted to get this fix in asap, will rebase later PRs on this PR).

LarryRuane commented on src/rpc/rawtransaction.cpp:
```diff
@@ -66,7 +66,9 @@ UniValue TxJoinSplitToJSON(const CTransaction& tx) {
         joinsplit.push_back(Pair("vpub_old", ValueFromAmount(jsdescription.vpub_old)));
+        joinsplit.push_back(Pair("vpub_oldZat", jsdescription.vpub_old));
         joinsplit.push_back(Pair("vpub_new", ValueFromAmount(jsdescription.vpub_new)));
+        joinsplit.push_back(Pair("vpub_newZat", jsdescription.vpub_new));
```
Same, fix problem reported here https://github.com/zcash/zcash/pull/3972#issuecomment-513510051 (but wanted to get this fix in asap, will rebase later PRs on this PR).

LarryRuane commented:
@str4d and @Eirik0, I think I've addressed all review comments as a separate commit, https://github.com/zcash/zcash/pull/3926/commits/38bcb0e8e74d6f7e6ad625ece2e3ecacb62bdc0a, so please take another look. 

I also added one more small commit, https://github.com/zcash/zcash/pull/3926/commits/7fa748490f30829579a0412e2c256f1f4e14d7b8, to fix a problem that was discovered in actual insight explorer testing. This bug was documented here https://github.com/zcash/zcash/pull/3972#issuecomment-513510051 but the fix belongs in an earlier PR, #3863, which is already merged, so I think it's best to get the fix into the earliest possible PR (the current one).

I think we should squash before the merge.

Eirik0 commented on src/rpc/misc.cpp:
```diff
@@ -483,6 +484,459 @@ UniValue setmocktime(const UniValue& params, bool fHelp)
+    bool includeChainInfo = false;
+    if (params[0].isObject()) {
+        UniValue chainInfo = find_value(params[0].get_obj(), "chainInfo");
+        if (!chainInfo.isNull()) {
```
This change causes it to throw an error if it is the wrong type rather than ignore it, which I think I suggested in an earlier comment.

str4d commented on src/rpc/misc.cpp:
```diff
@@ -712,8 +755,20 @@ static void getHeightRange(const UniValue& params, int& start, int& end)
         }
     }
+
+    if (start > chainActive.Height() || end > chainActive.Height()) {
```
`chainActive` requires a lock on `cs_main`. Either grab the lock here, add an `AssertLockHeld(cs_main)` to the start of this function, or move this check out to the individual RPC methods. Ideally whichever option minimises the necessary locking period, and keeps the information gathered from the various locked sections consistent.

str4d commented on src/test/rpc_tests.cpp:
```diff
@@ -346,4 +360,61 @@ BOOST_AUTO_TEST_CASE(rpc_getnetworksolps)
+        "Run './zcash-cli help getaddresstxids' for instructions on how to enable this feature.");
+
+    fExperimentalMode = true;
+    fInsightExplorer = true;
```
Revert these at the end of the test (e.g. this test would fail if it were run immediately after itself, due to mutation of global state).

LarryRuane commented:
Thanks, @str4d, both problems should now be fixed.

LarryRuane commented:
@str4d, @Eirik0, there's yet another problem discovered by actual Insight use, https://github.com/zcash/zcash/pull/3972#issuecomment-515638085 (zero fees reported), and even though that testing was using a later PR branch, I suspect the bug exists in the current branch. So I'd like to see if I can get the fix into this PR, but I'm not sure what the fix will be yet (I don't know the cause). If you'd rather, I can fix in a later PR. I will update here as soon as I can.

LarryRuane commented:
@str4d, @Eirik0, I believe I have fixed the problem, please review latest commit 080fd80f1f329337814d35ea2009ac8ed4fe7ac4 (this is a field that is present in the hackworks patched zcash but I missed it in the merge).

str4d commented on src/rpc/misc.cpp:
```diff
@@ -898,18 +899,18 @@ UniValue getaddressdeltas(const UniValue& params, bool fHelp)
-    startInfo.push_back(Pair("hash", startIndex->GetBlockHash().GetHex()));
+    {
+        LOCK(cs_main);  // for chainActive
+        if (start > chainActive.Height() || end > chainActive.Height()) {
```
AFAICT this conditional will never be reached, because `getHeightRange()` (which is called in this function to set `start` and `end`) contains the same check. Not clear to me from this commit which check is now redundant (i.e. whether `getHeightRange()` is used elsewhere).

LarryRuane replied:
My thinking here is that the return value of `chainActive.Height()` may be different from when it was sampled from within the call to `getHeightRange()` since `cs_main` was released in the meantime. I wanted to make sure that `start` and `end`, when used as indices to `chainActive` immediately after this check:
```
startInfo.push_back(Pair("hash", chainActive[start]->GetBlockHash().GetHex()));
endInfo.push_back(Pair("hash", chainActive[end]->GetBlockHash().GetHex()));
```
are valid. Do I have an incorrect understanding?

str4d replied:
No, that's accurate. I was imprecise in my earlier comment: the check is redundant except if a reorg happens in parallel with this command executing (which is the only way that `chainActive.Height()` could decrease between the two checks). Even then, I'm not sure it could be triggered particularly easily, as the reorg itself will hold the main lock, and for most small reorgs the result is going to be at least the same height (in order for the reorg to be towards a chain with more work). But I agree that this eliminates the possibility, so I'm happy to leave it in.

LarryRuane commented:
Thanks, @str4d and @Eirik0, I just rebased and force-pushed (no changes). I was not able to fix that bug yet, so will do that in a later commit. I think this is ready for r+.

str4d commented:
@zkbot r+
